python-api/tests/integrationtests/baseclass.py

```diff
@@ -0,0 +1,412 @@
+
+        self.log.info("connecting Controller to Server")
+        self.controller.establish_connection()
+        self.serv.wait_for_output('registered: ')
+
+        assert self.controller.connection is not None
+
+    def new_test_video(self, pattern=None):
+        """Start a new Video-Testsource and wait until it's ready"""
+        self.serv.wait_for_output(':::3000')
+        self._sources.new_test_video(pattern=pattern)
+
+    def new_test_audio(self, freq=110, wave=AudioSrc.WAVE_SINE):
+        """Start a new Audio-Testsource and wait until it's ready"""
+        self.serv.wait_for_output(':::4000')
+        self._sources.new_test_audio(freq=freq, wave=wave)
+
+    def teardown_method(self, _):
+        """Tear down called automatically after every test_XXXX method."""
+        self.controller = None
+
+        # Kill all the sources
+        if self._sources is not None:
+            self.log.info("terminating Video-TestSource")
+            self._sources.terminate_video()
+
+            self.log.info("terminating Audio-TestSource")
+            self._sources.terminate_audio()
+        self._sources = None
+
+        if self.serv is not None:
+            self.log.info("terminating Server")
+            self.serv.terminate(cov=True)
+
+            if self.serv.proc:
+                poll = self.serv.proc.poll()
+                if poll == -11:
+                    self.log.error("Server exited with Segmentation Fault")
+                if poll != 0:
+                    self.log.error("Server exited Error Eode %s", poll)
+
+            log = open('server.log')
+            print(log.read())
+
+        self.serv = None
+
+
+class IntegrationTestbaseMainloop(IntegrationTestbase):
+    """Base class for integration tests that require a GLib-Mainloop.
+    Used for Tests that register and wait for signals.
+    """
+    def setup_method(self, method):
+        """Set up called automatically before every test_XXXX method."""
+        super(IntegrationTestbaseMainloop, self).setup_method(method)
+        self.mainloop = GLib.MainLoop()
+        self.quit_count = 0
+
+    def run_mainloop(self, timeout=5):
+        """Start the MainLoop, set Quit-Counter to Zero"""
+        self.quit_count = 0
+        GLib.timeout_add_seconds(timeout, self.quit_mainloop)
+        self.mainloop.run()
+
+    def quit_mainloop(self, *_):
+        """Quit the MainLoop, set Quit-Counter to Zero"""
+        self.mainloop.quit()
+        self.quit_count = 0
+
+    def quit_mainloop_after(self, call_count):
+        """Increment Quit-Counter, if it reaches call_count,
+        Quit the MainLoop"""
+        self.quit_count += 1
+        if self.quit_count == call_count:
+            self.quit_mainloop()
+
+
+class IntegrationTestbaseCompare(IntegrationTestbase):
+    """Base class for integration tests that require image capture
+    and comparison. Used for Video- and Audio-Tests.
+    """
+
+    def expect_frame(self, filename, port, turns=5, timeout=0.5):
```

Comment:
I have to let the `gst-launch` process run for a certain amount of time before starting to compare the output-frames.
Trying to compare them asynchronously is not reliable enough, especially in a slow environment like travis.
To if I let `gst-launch` run for, say, 5 seconds, the test can not succeed in under this amount of time.

Instead I start `gst-launch` up to 5x and have each run for 1s, so in the best case the test will succeed after 1s but it can take up to 5s if required. I'll document that in the code.